rotkehlchen/chain/bitcoin/utils.py

```diff
@@ -107,21 +188,91 @@ def pubkey_to_p2sh_p2wpkh_address(data: bytes) -> BTCAddress:
+    converted_bits = bech32.convertbits(witprog, 8, 5)
+    if converted_bits is None:
+        raise EncodingError('Could not derive Bech32 address from given public key')
+    ret = _bech32_encode([witver] + converted_bits, spec)
+    if _bech32_decode(ret) == (None, None):
+        raise EncodingError('Could not derive Bech32 address from given public key')
+    return BTCAddress(ret)
+
+
+def _tweak_public_key(data: bytes) -> bytes:
+    """Tweak a public key as defined by BIP-0086.
+
+    tweaked_pub_key = lift_x(pub_key.X()) + int(HashTapTweak(bytes(pub_key.X()))) * G
+
+    Code is taken from:
+    https://github.com/ebellocchia/bip_utils/blob/master/bip_utils/addr/P2TR_addr.py#L117
+    """
+    h = _hash_tap_tweak(data)
+    lift_x = _lift_x(data)
+    lhs = PublicKey.from_point(x=lift_x[0], y=lift_x[1])  # left hand side of the addition
+    rhs = PublicKey.from_point(x=INFINITY_COORDINATE[0], y=INFINITY_COORDINATE[1]).multiply(h)  # right hand side of the addition  # noqa: E501
+    result = lhs.combine([rhs])
+
+    result_x = result.point()[0]
+    return result_x.to_bytes(32, 'big')
+
+
+def _hash_tap_tweak(data: bytes) -> bytes:
+    """Compute the HashTapTweak of the specified public key.
+    Tagged hash = SHA256(SHA256(tag) || SHA256(tag) || data)
+
+    Code is taken from:
+    https://github.com/ebellocchia/bip_utils/blob/master/bip_utils/addr/P2TR_addr.py#L59
+    https://github.com/ebellocchia/bip_utils/blob/master/bip_utils/addr/P2TR_addr.py#L76
+    """
+    pubkey = PublicKey(data)
+    pubkey_x = pubkey.point()[0]  # the x coordinate of the point.
+    pubkey_x_bytes = pubkey_x.to_bytes(32, 'big')
+    return hashlib.sha256(TAP_TWEAK_SHA256 + TAP_TWEAK_SHA256 + pubkey_x_bytes).digest()
+
+
+def _lift_x(data: bytes) -> Tuple[int, int]:
+    """Implementation of the lift_x function as defined by BIP-0340.
+    It computes the point P for which P.X() = pub_key.X() and has_even_y(P).
+
+    Code is taken from:
+    https://github.com/ebellocchia/bip_utils/blob/master/bip_utils/addr/P2TR_addr.py#L92
+
+    May raise:
+    - ValueError if conditions are not met.
+    """
+    pubkey = PublicKey(data)
+    p = FIELD_SIZE
+    x = pubkey.point()[0]  # x coordinate of the point
+    if x >= p:
+        raise ValueError('Unable to compute lift_x point')
+    c = (pow(x, 3, p) + 7) % p
+    y = pow(c, (p + 1) // 4, p)
+    if c != pow(y, 2, p):
+        raise ValueError('Unable to compute lift_x point')
+
+    return (x, y if y % 2 == 0 else p - y)
+
+
+def _bech32_encode(data: List[int], spec: BitcoinSegwitEncoding) -> str:
+    combined = data + _bech32_create_checksum(data, spec)
+    return 'bc1' + ''.join([bech32.CHARSET[d] for d in combined])
+
 
-    return BTCAddress(result)
+def _bech32_create_checksum(data: List[int], spec: BitcoinSegwitEncoding) -> list:
+    values = bech32.bech32_hrp_expand('bc') + data
+    const = BECH32M_CONST if spec == BitcoinSegwitEncoding.BECH32M else 1
```

Comment:
If I understand correctly, the only difference between this method and `bech32.create_checksum` is the `BECH32M_CONST`. So I'm just curious, what is it?

Reply:
It's tricky to explain as I just got to know about this two days ago but this [comment should help](https://github.com/sipa/bech32/issues/51#issuecomment-554885190).

TLDR: Basically, it's the attempt at rectifying the woes of `BIP-173`

Sorry, you're probably exhausted from hearing BIP BIP BIP 😄

Reply:
Ha-ha, thanks, got it